text_generation/causal_lm/cpp/speculative_decoding_lm.cpp

```diff
@@ -0,0 +1,294 @@
+
+            forward_key_values(draft_model);  // forward KV cache from Result outputs to Parameter inputs
+            draft_model.infer();
+
+            auto draft_logits = draft_model.get_tensor("logits").data<float>();
+            int64_t arg_max_token = std::max_element(draft_logits, draft_logits + vocab_size) - draft_logits;
+            next_token = arg_max_token;
+            draft_tokens.emplace_back(arg_max_token);
+        }
+
+        // For the main network, K tokens will be fed at once in a single infer request.
+        input_ids.set_shape({BATCH_SIZE, K});
+        // Set the first token for the main model to be the same as for the draft model.
+        input_ids.data<int64_t>()[0] = first_token;
+        for (int i = 0; i < K - 1; i++)
+            input_ids.data<int64_t>()[i + 1] = draft_tokens[i];
+
+        attention_mask.set_shape({BATCH_SIZE, seq_len + K});
+        std::fill_n(attention_mask.data<int64_t>(), attention_mask.get_size(), 1);
+
+        position_ids.set_shape({BATCH_SIZE, K});
+        std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), seq_len);
+
+        forward_key_values(main_model);
+        main_model.infer();
+
+        data_logits = logits.data<float>();  // [BATCH_SIZE, K, vocab_size]
+        size_t disagree_idx = K - 1;
+        // Iterate through the predicted tokens from the main model and compare them with draft predictions.
+        // In the worst-case scenario (disagreement at the beginning), iter will increase by 1.
+        // In the best-case scenario, all elements match, and K predicted tokens will be taken. 
+        for (size_t i = 0; i < K; i++) {
+            auto start = data_logits + vocab_size * i;
+            auto stop = data_logits + vocab_size * (i + 1);
+            next_token = std::max_element(start, stop) - start;
+            text_streamer.put(next_token);
+
+            disagree_idx = i;                
+            if (next_token != draft_tokens[i])
+                break;
+        }
+
+        // After the inference request, key/values have shape [BATCH_SIZE, seq_len + K, vocab_size].
+        // Increment the sequence length by the number of matched tokens, and
+        // trim the KV cache to match the new sequence length.
+        seq_len += disagree_idx + 1;
+        update_kv_cache(draft_model, SEQ_LEN_AXIS, seq_len);
+        update_kv_cache(main_model, SEQ_LEN_AXIS, seq_len);
+        
+        draft_tokens.clear();
+        first_token = next_token;
+    }
+    text_streamer.end();
+    // Model is stateful which means that context (kv-cache) which belongs to a particular
+    // text sequence is accumulated inside the model during the generation loop above.
+    // This context should be reset before processing the next text sequence.
+    // While it is not required to reset context in this sample as only one sequence is processed,
+    // it is called for education purposes:
+    draft_model.reset_state();
```

Comment:
looks like it's not required, because we don't work with stateful model

Reply:
removed